### ISSUE TITLE
bug in np.arange? 

### ISSUE THREAD

**Issue by jgerlach93** [numpy/numpy#19552]
Let me show you the issue directly: 

### Reproducing code example:

np.arange(2114.8,2114.8+.05,.05)  produces array([2114.8 , 2114.85])

np.arange(2114.8,2114.8+1.,1.) produces  array([2114.8]) 

dont understand why they return arrays of different lengths.

**Comment by rossbar** [comment]
This is a common "gotcha" with arange - the reasons behind it are addressed in the [docstring](https://numpy.org/devdocs/reference/generated/numpy.arange.html#numpy.arange), particularly in the `Warnings` section. Hopefully that addresses your question and provides guidance re: the preferred alternative (np.linspace). If not, ideas for improvements are welcome

**Comment by jgerlach93** [comment]
thanks for the response. maybe i am very picky, but i find that it would be more consistent when it works like with 0.05 when i specify 1. (as a float) and the other way when i specify 1 as an int. But maybe this would also cause some other confusion in return.

**Comment by charris** [comment]
The problem is that `.05` is not equal to `.05` after conversion to (base two) floating point.

**Comment by bashtage** [comment]
This is literally the most common question on the issue tracker.  Should ping this or one of the other variants.

**Comment by melissawm** [comment]
I think this means the documentation is insufficient. Maybe the warning is too far down the docstring and we should mention the gotcha earlier?

**Comment by eric-wieser** [comment]
> Should ping this or one of the other variants.

I created https://github.com/numpy/numpy/projects/11 a while ago to track this.

**Comment by melissawm** [comment]
Linking related issues for reference: #19670, #18212, #18563

**Comment by rossbar** [comment]
> The current docs have a clear warning. Should this be closed?

Thanks for the ping @soegaard - I agree; since the issue has been tagged in the `arange` project and there's nothing specifically actionable here I will close this.

**Comment by HotDog702** [comment]
Similar issue, that always get me stuck:
`X = np.arange(L, R, step=1/4000)`
and
`X = np.arange(0, R-L, step=1/4000)+L`
for L = 1710922191.203908
and R = 1710922252.1371984
returns different output on the right end, namely 99 last points in X greater than R for the first variant !!! How it can ever been???
May be internally always use the second code?

**Comment by jakevdp** [comment]
@HotDog702 What you're seeing is related to this:
```python
>>> 0.1 + 0.2 == 0.3
False
```
https://stackoverflow.com/q/588004 has a pretty good discussion of why expressing the "same" computation in two different ways can lead to different results with floating point math.

**Comment by HotDog702** [comment]
@jakevdp Yes, I know that. But the thing that makes me crazy is that the result is **out of defined range** L-R. 
OK, I can understand the last point >R, but 99 points!!! I use the above code to define the new interpolation points **with certain step** and **within defined range**. So such behavior is really annoying.
I saw the advice to use the `np.linspace`, but this is also the dance around the problem
